File: build_forward_index.cpp
#include <iostream>
#include <fstream>
#include <filesystem>
#include <map>
#include <vector>
#include <sstream>
#include <algorithm>

using namespace std;
namespace fs = std::filesystem;

string DOCS_PATH = "D:/Dataset/Cleaned1/";
string OUTPUT_PATH = "D:/DSA-Project/search_engine/backend/app/indexing/output/";
string LEXICON_PATH = "D:/DSA-Project/search_engine/backend/app/indexing/output/lexicon.csv";

vector<string> tokenize(const string &text) {
    vector<string> tokens;
    string word;

    for (char c : text) {
        if (isalnum(c)) word += tolower(c);
        else {
            if (!word.empty()) {
                tokens.push_back(word);
                word.clear();
            }
        }
    }
    if (!word.empty()) tokens.push_back(word);

    return tokens;
}

map<string, int> loadLexicon() {
    map<string, int> lex;
    ifstream in(LEXICON_PATH);
    string word, line;
    getline(in, line);

    while (getline(in, line)) {
        stringstream ss(line);
        string w, wid;
        getline(ss, w, ',');
        getline(ss, wid, ',');
        lex[w] = stoi(wid);
    }

    return lex;
}

int main() {
    map<string, int> lexicon = loadLexicon();
    map<int, vector<int>> forwardIndex;
    int docID = 1;

    for (const auto &file : fs::directory_iterator(DOCS_PATH)) {
        if (file.path().extension() == ".txt") {
            ifstream in(file.path());
            stringstream buffer;
            buffer << in.rdbuf();
            string text = buffer.str();
            vector<string> tokens = tokenize(text);

            vector<int> wordIDs;
            for (auto &t : tokens) {
                if (lexicon.count(t)) {
                    wordIDs.push_back(lexicon[t]);
                }
            }
            forwardIndex[docID++] = wordIDs;
        }
    }

    ofstream out(OUTPUT_PATH + "forward_index.csv");
    out << "docID,wordIDs\n";

    for (auto &p : forwardIndex) {
        out << p.first << ",";
        for (int id : p.second) out << id << " ";
        out << "\n";
    }

    cout << "forward_index.csv generated!" << endl;
    return 0;
}

File: build_forwardindex.py
import os, json, gzip, re, csv

JSON_PATH = "D:/DATA/"
LEXICON = "D:/DSA-Project/search_engine/backend/app/indexing/output/lexicon.csv"
OUTPUT = "D:/DSA-Project/search_engine/backend/app/indexing/output/forward_index.csv"

def tokenize(t):
    return re.findall(r"[a-z0-9]+", t.lower())

# load lexicon
lex = {}
with open(LEXICON, encoding="utf-8") as f:
    next(f)
    for line in f:
        w, i = line.strip().split(",")
        lex[w] = int(i)

with open(OUTPUT, "w", newline="", encoding="utf-8") as out:
    writer = csv.writer(out)
    writer.writerow(["docID", "wordIDs"])

    doc_id = 1

    for file in os.listdir(JSON_PATH):
        path = os.path.join(JSON_PATH, file)
        f = gzip.open(path, "rt", encoding="utf-8") if file.endswith(".gz") else open(path, encoding="utf-8")
        data = json.load(f)

        text = ""
        for sec in data.get("abstract", []):
            text += sec.get("text", "") + " "
        for sec in data.get("body_text", []):
            text += sec.get("text", "") + " "

        ids = [str(lex[t]) for t in tokenize(text) if t in lex]
        writer.writerow([doc_id, " ".join(ids)])

        doc_id += 1

print("Forward index created")
